torch/quantization/convert_modules.py

```diff
@@ -0,0 +1,221 @@
+
+def addQuantDeQuantCall(base_fn):
+    r"""Insert calls to quant and dequant around base_fn calls
+
+    Given a `base_fn` function, insert a call to `quant` before the function
+    call which is supposed to quantize the input and a call to `dequant` after
+    the function call which will dequantize the output of the function.
+
+    Args:
+        base_fn: base function we want to call
+
+    Returns:
+        new_fn: a modified function which calls quant -> base_bn -> dequant
+    """
+    def new_fn(x):
+        x = base_fn.__self__.quant(x)
+        x = base_fn(x)
+        x = base_fn.__self__.dequant(x)
+        return x
+    return new_fn
+
+def addQuantDeQuantModule(mod, qConfig):
+    r"""Add Quantize and DeQuantize to the input module and modify forward function
+
+    Args:
+        mod: module to be changed
+        qConfig: quantization configuration
+
+    """
+    mod.add_module('quant', AbstractQuant(qConfig))
+    mod.add_module('dequant', AbstractDeQuant())
+    mod.forward = addQuantDeQuantCall(mod.forward)
+
+def get_config_key(name, qConfigDict):
+    if name in qConfigDict:
+        return name
+    elif name:
+        parent = '.'.join(name.split('.')[:-1])
+        return get_config_key(parent, qConfigDict)
+    else:
+        return '' if '' in qConfigDict else None
+
+def getModule(modInstance, name):
+    splits = name.split('.')
+    mod = modInstance._modules[splits[0]]
+    for i in range(1, len(splits)):
+        mod = mod._modules[splits[i]]
+    return mod
+
+def addQuantDeQuant(modInstance, qConfigDict):
+    r"""Add AbstractQuant and AbstractDeQuant module and modify forward function
+    to call quant dequant accordign to qConfigDict
+    """
+    mod_key_list = []
+    for name, mod in modInstance.named_modules():
+        dict_key = get_config_key(name, qConfigDict)
+        if dict_key is not None:
+            mod_key_list.append(dict_key)
+
+    for name in set(mod_key_list):
+        if name is not '':
+            addQuantDeQuantModule(getModule(modInstance, name), qConfigDict[name])
+        else:
+            addQuantDeQuantModule(modInstance, qConfigDict[name])
+
+def prepare(modInstance, qConfigDict):
+    r"""Transform a module to a quantized module according to qConfigDict,
+    it adds observer and abstract quant dequant modules to the module. and
+    changes the forward method to invoke them
+
+    Args:
+        modInstance: instance of the module we want to transform
+        qCofnigDict: dictionary that maps from name of submodule to quantization
+                     configuration
+    """
+    addObserver(modInstance, qConfigDict)
+    addQuantDeQuant(modInstance, qConfigDict)
+
+class AbstractQuant(nn.Module):
+    r"""Abstract quantize module which will be replace to actual Quantize module
+    in swapModule function
+    """
+    def __init__(self, qconfig):
+        super(AbstractQuant, self).__init__()
+        self.add_module('observer', qconfig.activation())
```

Comment:
why do you have observer here if you already have an observer on parent module. Something doesn't add up here

Reply:
removed.

Reply:
actually we need this for input activations I think.